core/src/test/java/dk/alexandra/fresco/framework/network/async/TestAsyncNetwork.java

```diff
@@ -0,0 +1,316 @@
+package dk.alexandra.fresco.framework.network.async;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import dk.alexandra.fresco.framework.Party;
+import dk.alexandra.fresco.framework.configuration.NetworkConfiguration;
+import dk.alexandra.fresco.framework.configuration.NetworkConfigurationImpl;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.nio.channels.ServerSocketChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+import javax.net.ServerSocketFactory;
+import org.junit.Test;
+
+public class TestAsyncNetwork {
+
+  private List<Integer> getFreePorts(int no) {
```

Comment:
There is really not much guarantee that these ports are free. In particular if tests are run in parrallel, we definitely do not get free ports. Find a better strategy.